File: solutions/3LongestSubstringWithoutRepeatingCharacters/main.go
/*
Given a string s, find the length of the longest
substring without duplicate characters.
*/

func lengthOfLongestSubstring(s string) int {
	n := len(s)
	maxLength := 0
	lastIdx := map[byte]int{}
	start := 0
	for end := range n {
		currentChar := s[end]
		if lastIdx[currentChar] > start {
			start = lastIdx[currentChar]
		}
		if end-start+1 > maxLength {
			maxLength = end - start + 1
		}
		lastIdx[currentChar] = end + 1
	}
	return maxLength
}

File: solutions/3LongestSubstringWithoutRepeatingCharacters/main.py
'''
Given a string s, find the length of the longest
substring without duplicate characters.
'''

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        last_idx = {}
        max_length = 0
        start = 0
        for end in range(n):
            current_char = s[end]
            if current_char in last_idx and last_idx[current_char] > start:
                start = last_idx[current_char]
            max_length = max(max_length, end-start+1)
            last_idx[current_char] = end + 1
        return max_length